src/mender-update/standalone/states.cpp

```diff
@@ -0,0 +1,507 @@
+	poster.PostEvent(StateEvent::Success);
+}
+
+void ArtifactInstallState::OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) {
+	auto err = ctx.update_module->ArtifactInstall();
+	if (err != error::NoError) {
+		log::Error("Installation failed: " + err.String());
+		UpdateResult(ctx.result_and_error, {Result::Failed, err});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	UpdateResult(ctx.result_and_error, {Result::Installed, error::NoError});
+	poster.PostEvent(StateEvent::Success);
+}
+
+void RebootAndRollbackQueryState::OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) {
+	auto reboot = ctx.update_module->NeedsReboot();
+	if (!reboot) {
+		log::Error("Could not query for reboot: " + reboot.error().String());
+		UpdateResult(ctx.result_and_error, {Result::Failed, reboot.error()});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	if (reboot.value() != update_module::RebootAction::No) {
+		UpdateResult(ctx.result_and_error, {Result::RebootRequired, error::NoError});
+	}
+
+	auto rollback_support = ctx.update_module->SupportsRollback();
+	if (!rollback_support) {
+		log::Error("Could not query for rollback support: " + rollback_support.error().String());
+		UpdateResult(ctx.result_and_error, {Result::Failed, rollback_support.error()});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	if (rollback_support.value()) {
+		poster.PostEvent(StateEvent::NeedsInteraction);
+		return;
+	}
+
+	cout << "Update Module doesn't support rollback. Committing immediately." << endl;
+	poster.PostEvent(StateEvent::Success);
+}
+
+void ArtifactCommitState::OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) {
+	auto err = ctx.update_module->ArtifactCommit();
+	if (err != error::NoError) {
+		log::Error("Commit failed: " + err.String());
+		UpdateResult(ctx.result_and_error, {Result::Failed, err});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	UpdateResult(ctx.result_and_error, {Result::Committed, error::NoError});
+	poster.PostEvent(StateEvent::Success);
+}
+
+void RollbackQueryState::OnEnter(Context &ctx, sm::EventPoster<StateEvent> &poster) {
+	auto rollback_support = ctx.update_module->SupportsRollback();
+	if (!rollback_support) {
+		log::Error("Could not query for rollback support: " + rollback_support.error().String());
+		UpdateResult(
+			ctx.result_and_error,
+			{Result::Failed | Result::RollbackFailed, rollback_support.error()});
+		poster.PostEvent(StateEvent::Failure);
+		return;
+	}
+
+	if (not rollback_support.value()) {
+		bool already_failed = ResultContains(ctx.result_and_error.result, Result::Failed);
+		UpdateResult(ctx.result_and_error, {Result::Failed | Result::NoRollback, error::NoError});
+		if (already_failed) {
+			poster.PostEvent(StateEvent::NothingToDo);
+		} else {
+			// If it hadn't failed already, it's because the user asked for the rollback
+			// explicitly. In this case bail out instead of continuing.
```

Comment:
I don't understand this comment. Are you referring to a user call `mender-update rollback` where the update module has no support for actual rollback?

Reply:
Exactly!